velox/core/Expressions.h

```diff
@@ -368,6 +371,9 @@ class ConcatTypedExpr : public ITypedExpr {
     if (!casted) {
       return false;
     }
+    if (casted->type() != this->type()) {
```

Comment:
would it make sense to generalize checks on (name, type, inputs) for all Expression and have that check in ITypedExpr ?

Reply:
Good question. I added type check for LambdaTypedExpr and FieldAccessTypedExpr too. But I found we cannot add type check for InputTypedExpr because it fails ExprTest.rewriteInputs. I think the problem is that InputTypedExpr::rewriteInputNames() still create a TypedExprPtr with the original type rather than the mapped type. This should ideally be fixed but in a separate diff.